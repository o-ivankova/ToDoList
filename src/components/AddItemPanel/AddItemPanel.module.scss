.addItemForm {
  border-style: groove;
  border: none;
  margin-bottom: 20px;
}
