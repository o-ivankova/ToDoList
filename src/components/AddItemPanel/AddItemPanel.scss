$focusbordercolor: #ffc2bd;

.add-item-panel {
  margin-bottom: 20px;
}

.form-control {
  border-style: groove;
  border: 1px solid lightgrey;
}

input:focus, input.form-control:focus {
    border: none;
    outline: 1px solid $focusbordercolor;
    outline-width: 1px !important;
    box-shadow: none;
    -moz-box-shadow: none;
    -webkit-box-shadow: none;
}
