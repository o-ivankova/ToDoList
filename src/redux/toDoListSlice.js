import { createSlice } from "@reduxjs/toolkit";
import storage from "redux-persist/lib/storage";
import { persistReducer } from "redux-persist";

const initialState = {
  elements: [],
  elementId: 0,
  filter: "all",
};

export const toDoListSlice = createSlice({
  name: "toDoList",
  initialState,
  reducers: {
    addElement: (state, action) => {
      const input = action.payload;

      if (input !== "") {
        state.elementId += 1;

        const newElement = {
          content: input,
          completed: false,
          id: state.elementId,
        };

        state.elements.push(newElement);
      }
    },

    deleteElement: (state, action) => {
      const id = action.payload;
      state.elements = state.elements.filter((el) => el.id !== id);
    },

    deleteAllCompleted: (state) => {
      state.elements = state.elements.filter((el) => !el.completed);
    },

    checkCompleted: (state, action) => {
      const id = action.payload;
      const index = state.elements.findIndex((el) => el.id === id);
      state.elements[index].completed = !state.elements[index].completed;
    },

    changeFilter: (state, action) => {
      const filter = action.payload;
      state.filter = filter;
    },

    reorderElements: (state, action) => {
      const result = action.payload;
      if (!result.destination) return;

      const [reorderedEl] = state.elements.splice(result.source.index, 1);
      state.elements.splice(result.destination.index, 0, reorderedEl);
    },
  },
});

export const {
  addElement,
  deleteElement,
  checkCompleted,
  deleteAllCompleted,
  changeFilter,
  reorderElements,
} = toDoListSlice.actions;

const persistConfig = {
  key: "root",
  storage,
};

const { reducer } = toDoListSlice;

export const persistedReducer = persistReducer(persistConfig, reducer);

// export default reducer;
