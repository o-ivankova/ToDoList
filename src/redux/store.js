import { configureStore } from "@reduxjs/toolkit";
import {persistedReducer} from "./toDoListSlice";
import { persistStore } from 'redux-persist'

export const store = configureStore({
  reducer: {
    toDoList: persistedReducer,
  },
});

export const persistor = persistStore(store);