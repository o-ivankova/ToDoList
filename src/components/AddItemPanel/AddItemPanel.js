import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { addElement } from "../../redux/toDoListSlice";
import TextInput from "./components/TextInput";
import styles from "./AddItemPanel.module.scss";

const AddItemPanel = () => {
  const dispatch = useDispatch();

  const [input, setInput] = useState("");

  const onInputChange = (e) => {
    setInput(e.target.value);
  };

  const onSubmit = (e) => {
    e.preventDefault();
    dispatch(addElement(input));
    setInput("");
  };

  return (
      <form className={styles.addItemForm} onSubmit={onSubmit} data-testid="addItemForm">
        <TextInput onInputChange={onInputChange} input={input}/>
      </form>
  );
};

export default AddItemPanel;
