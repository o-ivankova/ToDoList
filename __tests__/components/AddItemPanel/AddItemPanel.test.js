import React from "react";
import "@testing-library/jest-dom";
import { jest, expect, test, describe, afterAll } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import AddItemPanel from "../../../src/components/AddItemPanel/AddItemPanel";
import { Provider } from "react-redux";
import { act } from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import { setupStore } from "../../../src/redux/store";
import renderer from "react-test-renderer";

const getMockStore = () => {
  const preloadedState = {
    toDoList: {
      filter: "all",
      elements: [],
    },
  };

  return setupStore(preloadedState);
}

describe("AddItemPanel tests", () => {
  test("component is rendered", () => {
    render(
      <Provider store={getMockStore()}>
        <AddItemPanel />
      </Provider>
    );

    screen.debug();
    expect(screen.getByTestId("addItemForm")).toBeInTheDocument();
    expect(screen.getByTestId("input")).toBeInTheDocument();
  });

  test("user input is visible", () => {
    render(
      <Provider store={getMockStore()}>
        <AddItemPanel />
      </Provider>
    );

    const userInput = "Buy milk";
    act(() => {
      userEvent.type(screen.getByRole("textbox"), userInput);
    });

    expect(screen.getByTestId("input").value).toBe(userInput);
  });

  test("input is saved on submit", () => {
    render(
      <Provider store={getMockStore()}>
        <AddItemPanel />
      </Provider>
    );

    const userInput = "Buy milk{enter}";
    act(() => {
      userEvent.type(screen.getByRole("textbox"), userInput);
    });

    const savedItem = JSON.parse(localStorage.getItem("toDoElements"));
    expect(savedItem.some((e) => e.content === "Buy milk"));
  });

  test("input is not saved on enter with no text", () => {
    render(
      <Provider store={getMockStore()}>
        <AddItemPanel />
      </Provider>
    );
    const initialSavedItem = localStorage.getItem("toDoElements");

    const userInput = "{enter}";
    act(() => {
      userEvent.type(screen.getByRole("textbox"), userInput);
    });

    const savedItemAfterInput = localStorage.getItem("toDoElements");
    expect(savedItemAfterInput).toEqual(initialSavedItem);
  });

  test("snapshot test", () => {
    const preloadedState = {
      toDoList: {
        elements: [
          { id: 1, content: "test1", completed: true },
          { id: 2, content: "test2", completed: true },
          { id: 3, content: "test3", completed: false },
        ],
      },
    };
    const mockStore = setupStore(preloadedState);

    const domTree = renderer
      .create(
        <Provider store={mockStore}>
          <AddItemPanel />
        </Provider>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });

  afterAll(() => {
    jest.clearAllMocks();
  }); 
});
