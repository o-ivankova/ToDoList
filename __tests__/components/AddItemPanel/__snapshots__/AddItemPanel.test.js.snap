// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddItemPanel tests snapshot test 1`] = `
<form
  className="addItemForm"
  data-testid="addItemForm"
  onSubmit={[Function]}
>
  <input
    className="addItemInput"
    data-testid="input"
    onChange={[Function]}
    placeholder="Type your to do"
    type="text"
    value=""
  />
</form>
`;
